vllm/core/scheduler.py

```diff
@@ -1306,6 +1324,39 @@ def _allocate_and_set_running(self, seq_group: SequenceGroup) -> None:
         for seq in seq_group.get_seqs(status=SequenceStatus.WAITING):
             seq.status = SequenceStatus.RUNNING
 
+    def _find_last_cached_block_id(self, seq_id: int) -> int:
+        """Finds the last block in the list with a non-None cached content hash."""
+        blocks = self.block_manager.block_tables[seq_id]._blocks
+        for block in reversed(blocks):
+            if block._cached_content_hash is not None:
+                return block.block_id
+        return None
+
+    def _add_pinned_caching_seq(self, seq_group: SequenceGroup) -> None:
+        running_seqs = seq_group.get_seqs(status=SequenceStatus.RUNNING)
+
+        if not running_seqs:
+            return
+
+        # NOTE: Here we assume that all sequences in the group have the same
+        # prompt.
+        seq = running_seqs[0]
+        if seq.pinned_caching:
+            device = Device.GPU
+            last_cached_block_id = self._find_last_cached_block_id(seq.seq_id)
+            num_blocks = self.block_manager.block_tables[seq.seq_id]
+
+            if last_cached_block_id is not None:
+                block_tracker = self.block_manager.block_allocator._allocators[device]._block_tracker[last_cached_block_id]
+                self.pinned_caching_manager.add_ttl_expiration_seq(seq, num_blocks, block_tracker)
```

Comment:
I have only implemented TTL expiration(which is like Anthropic prompt caching style) for now, but other expiration schemes can also be supported. I am considering linking each expiration scheme to API options.